### PR TITLE
refactor cluster controller

### LINKED ISSUE #3234: The tower server shouldn't update the cluster.status
**What's it about?**
Currently, when cluster.spec.connection.type is proxy, the tower server updates the cluster.status filed, and the cluster controller determines if the cluster is ready according to the status.

**What's the reason why we need it?**
Actually, the tower server shouldn't update the cluster.status filed.
The cluster controller can try to connect to the proxy service, determines if the cluster is ready by the response HTTP code.
This proposal can reduce the complexity between tower server and clsuter-controller.
/area multicluster
/kind feature-request

### PR DESCRIPTION
**Pull request by yuswift**
Signed-off-by: yuswift <yuswiftli@yunify.com>

**What type of PR is this?**
/kind design
**What this PR does / why we need it**:
Reduce the complexity between tower server and clsuter-controller. Remove the `port allocation` `proxy creation` `token generation` steps, add `cluster ready detection`  step.
**Which issue(s) this PR fixes**:
Fixes #3234

### CODE REVIEW

File: pkg/controller/cluster/cluster_controller.go [hunk 1]
```diff
@@ -348,6 +345,80 @@ func (c *clusterController) reconcileHostCluster() error {
...
+			klog.Error(err)
+			continue
+		}
+
+		config, err := clientConfig.ClientConfig()
+		if err != nil {
+			klog.Error(err)
+			continue
+		}
+		config.Timeout = 10 * time.Second
```
Comment (zryfish) [inline pkg/controller/cluster/cluster_controller.go:370]: What if there are lots of clusters, saying 50, and each cluster takes 9s to finish probing, that would be 450s, > resyncPeriod.

File: pkg/simple/client/multicluster/options.go [hunk 2]
```diff
@@ -79,5 +79,5 @@ func (o *Options) AddFlags(fs *pflag.FlagSet, s *Options) {
 		"This field is used when generating deployment yaml for agent.")
 
 	fs.DurationVar(&o.ClusterControllerResyncSecond, "cluster-controller-resync-second", s.ClusterControllerResyncSecond,
-		"Cluster controller resync second to sync cluster resource.")
+		"Cluster controller resync second to sync cluster resource. e.g. 30s 60s 120s...")
```
Comment (zryfish) [inline pkg/simple/client/multicluster/options.go:82]: better to start `2m, 5m, 10m`, small resync period increases load

File: pkg/controller/cluster/cluster_controller.go [hunk 3]
```diff
@@ -348,6 +345,80 @@ func (c *clusterController) reconcileHostCluster() error {
 	return err
 }
 
+func (c *clusterController) judgeIfClusterIsReady() error {
```
[the remark this answers is not in the document]
Comment (yuswift) [reply pkg/controller/cluster/cluster_controller.go:348]: Agreed.

File: pkg/controller/cluster/cluster_controller.go (hunk 1 above)
Comment (yuswift) [reply pkg/controller/cluster/cluster_controller.go:370]: 10 seconds seem to be too long. But the case you are talking about is very rare. It's unlikely that every cluster connection takes 9s. In most cases, one connection takes several ms. How about change the timeout to 3s?

File: pkg/simple/client/multicluster/options.go (hunk 2 above)
Comment (yuswift) [reply pkg/simple/client/multicluster/options.go:82]: Agree with that. I will update the comment.

File: pkg/controller/cluster/cluster_controller.go (hunk 1 above)
Comment (zryfish) [reply pkg/controller/cluster/cluster_controller.go:370]: If there are network issues on the node where the `ks-controller-manager` pod residing, it's possible.

Comment (zryfish) [reply pkg/controller/cluster/cluster_controller.go:370]: What I did before put cluster back to working queue every resyncPeriod, and check its readiness on main sync loop.

Comment (yuswift) [reply pkg/controller/cluster/cluster_controller.go:370]: I didn't see ` put cluster back to working queue every resyncPeriod`, but I saw `check its readiness on main sync loop.`. I think we don't need to `put cluster back to working queue every resyncPeriod` manually, the cluster informer does that automatically. The reason I check the cluster readiness separately is to check all of the cluster readiness, not only proxy connection. What you did before only checks the proxy cluster if has `agent availably` status, then updates the cluster status to ready or not. By using kubeconfig, I think it's more reliable.

- [ ]

Comment (yuswift) [reply pkg/controller/cluster/cluster_controller.go]: If put in the main sync loop, once the cluster is created/updated/deleted, the check will be performed, which may be too frequent. On the other hand, the check may be too long and will affect the sync loop. what is your suggestion?

Comment (zryfish) [reply pkg/controller/cluster/cluster_controller.go]: That's true, but we need to update `cluster.status.configz` every resyncPeriod too. So I suggest make `config.timeout` shorter, and probe in the main loop.

Comment (yuswift) [reply pkg/controller/cluster/cluster_controller.go]: Currently, we update the `cluster.status.configz` every resyncPeriod at the end main loop. The update doesn't change.

Comment (zryfish) [reply pkg/controller/cluster/cluster_controller.go]: OK, better to make `config.timeout` shorter

Comment (yuswift) [reply pkg/controller/cluster/cluster_controller.go]: `config.Timeout` has been set as 3s by default. We can merge this pr now.

**Pull request merged**